ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -310,100 +311,126 @@ forkBlockProduction
+        -- within @k@ blocks from the tip anymore. It's a theoretical possibility,
+        -- but very unlikely. When it happens, we simply give up on the chance
+        -- to produce a block.
+        extLedger <- do
+          mExtLedger <- lift $ ChainDB.getPastLedger chainDB prevPoint
+          case mExtLedger of
+            Just l  -> return l
+            Nothing -> do
+              trace $ TraceNoLedgerState currentSlot prevPoint
+              exitEarly
+        let ledger = ledgerState extLedger
+
+        -- Check if we are the leader
+        proof <-
+          case anachronisticProtocolLedgerView cfg ledger (At currentSlot) of
+            Right ledgerView -> do
+              mIsLeader <- lift $ atomically $ runProtocol varDRG $
+                checkIsLeader
+                  cfg
+                  currentSlot
+                  ledgerView
+                  (ouroborosChainState extLedger)
+              case mIsLeader of
+                Just p  -> return p
+                Nothing ->
+                  -- We're not the leader. This is the normal case, and so we
+                  -- don't need to trace a message.
+                  exitEarly
+            Left err -> do
+              -- There are so many empty slots between the tip of our chain and
+              -- the current slot that we cannot even get an accurate ledger
+              -- view anymore. This is indicative of a serious problem: we are
+              -- not receiving blocks. It is /possible/ it's just due to our
+              -- network connectivity, and we might still get these blocks at
+              -- some point; but we certainly can't produce a block of our own.
+              trace $ TraceCouldNotForge currentSlot err
+              exitEarly
+
+        -- At this point we have established that we are indeed slot leader
+        trace $ TraceForgeAboutToLead currentSlot
+
+        -- Get a snapshot of the mempool that is consistent with the ledger
+        --
+        -- NOTE: It is possible that due to adoption of new blocks the
+        -- /current/ ledger will have changed. This doesn't matter: we will
+        -- produce a block that fits onto the ledger we got above; if the
+        -- ledger in the meantime changes, the block we produce here may or
+        -- may not be adopted, but it won't be invalid.
+        mempoolSnapshot <- lift $ atomically $ getSnapshotFor
+                                                 mempool
+                                                 (TxsForBlockInSlot currentSlot)
+                                                 (ledgerState extLedger)
+        let blockEncOverhead = nodeBlockEncodingOverhead ledger
+            maxBlockBodySize = case maxBlockSizeOverride of
+              NoOverride            -> nodeMaxBlockSize ledger - blockEncOverhead
+              MaxBlockSize mbs      -> mbs - blockEncOverhead
+              MaxBlockBodySize mbbs -> mbbs
```

Comment:
Current configuration (max block size) could become out of date with max block size on the ledger; should take a `min` here.

Reply:
(Make it part of TraceForgeEvent)